hathor/event/websocket/factory.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Set
+
+from autobahn.twisted.websocket import WebSocketServerFactory
+from pydantic import ValidationError
+from structlog import get_logger
+
+from hathor.event import BaseEvent
+from hathor.event.storage import EventStorage
+from hathor.event.websocket.protocol import EventWebsocketProtocol
+from hathor.event.websocket.request import StreamRequest
+from hathor.event.websocket.response import BadRequestResponse, EventResponse, EventWebSocketNotRunningResponse
+from hathor.util import json_dumpb
+
+logger = get_logger()
+
+
+class EventWebsocketFactory(WebSocketServerFactory):
+    """ Websocket that will handle events
+    """
+
+    protocol = EventWebsocketProtocol
+    _is_running = False
+    _latest_event_id: Optional[int] = None
+
+    def __init__(self, event_storage: EventStorage):
+        super().__init__()
+        self.log = logger.new()
+        self._event_storage = event_storage
+        self._connections: Set[EventWebsocketProtocol] = set()
+
+        latest_event = self._event_storage.get_last_event()
+
+        if latest_event is not None:
+            self._latest_event_id = latest_event.id
+
+    def start(self):
+        """Start the WebSocket server. Required to be able to send events."""
+        self._is_running = True
+
+    def stop(self):
+        """Stop the WebSocket server. No events can be sent."""
+        self._is_running = False
+
+        for connection in self._connections:
+            connection.sendClose()
+
+        self._connections.clear()
+
+    def broadcast_event(self, event: BaseEvent) -> None:
+        """Broadcast the event to each registered client."""
+        self._latest_event_id = event.id
+
+        for connection in self._connections:
```

Comment:
I feel it's reasonable to assume `self._connections` is small enough to lock the event loop without issues. Just adding this note to confirm.

Reply:
I could just change it to use Twisted's `callLater()` method as I did in https://github.com/HathorNetwork/hathor-core/pull/522#discussion_r1099319225. Do you think it would be better?

Reply:
I don't think it would be useful. In fact, each use case will run their own full node to access this integration API. So it shouldn't have too many simultaneous connections.

Reply:
I'm almost certain that `self._send_event_to_connection` is non-blocking, or at least it _should_ be non-blocking. So the looping is not an issue, I think it's even expected that simple _broadcasts_ are implemented this way.

Reply:
Thanks @jansegre .